tools/authenticator.go

```diff
@@ -0,0 +1,69 @@
+package tools
+
+import (
+	"net/http"
+	"github.com/GoKillers/libsodium-go/cryptobox"
+	"github.com/GoKillers/libsodium-go/randombytes"
+	"errors"
+	b64 "encoding/base64"
+	"encoding/hex"
+)
+
+type AuthError struct {//todo
+	Url string
+	ApiKey string
+	Err error
+}
+
+type Authenticator struct{
+	publicKeys map[string]string
+}
+
+//func NewAuthenticator() *Authenticator{
+//
+//}
+
+func (authenticator Authenticator) Authenticate(url string, apiKey string, headers []http.Header){
+	if serverPk, ok := authenticator.publicKeys[url]; !ok{
+		serverPk = authenticator.GetServerPK(url)
+		authenticator.publicKeys[url] = serverPk
```

Comment:
I don't this that `map[string]string` is safe for concurrent usage.
On Python, there is no concurrently involved, and on C#, ISTR we use a concurrent dictionary